adds features to implement MCP server responsibilities in agentgateway

rinormaloku opened the pull request:
This PR adds the options to configure MCP Server responsibilities in the agentgateway.

- Returning 401 and WWW-Authentication header when the requests are unauthenticated
- Custom authorization server with jwks location
- Adds more metadata to the resource metadata endpoint.

Additionally, it adds a test authorization server to try out the flows

See the example under `examples/mcp-authentication/README.md` for quickly testing out the changes.

rinormaloku commented on ui/src/lib/backend-hooks.ts:
```diff
@@ -57,7 +57,7 @@ export const useBackendData = () => {
       // Auto-expand binds with backends
       const bindsWithBackends = new Set<number>();
-      backends.forEach(({ bind }) => bindsWithBackends.add(bind.port));
+      backends?.forEach(({ bind }) => bindsWithBackends.add(bind.port));
```
what is the benefit of one vs the other?

amivanoff commented on crates/agentgateway/src/mcp/sse.rs:
```diff
@@ -230,14 +246,32 @@ pub struct McpTarget {
+	}
+
+	async fn protected_resource_metadata(&self, req: Request, auth: McpAuthentication) -> Response {
 		let new_uri = Self::get_redirect_url(&req, "/.well-known/oauth-protected-resource");
```
It looks like `/.well-known/oauth-protected-resource` is not a suffix (line 311), but a prefix in URL Path. So it ends up with errors like this `http://localhost:3003/.well-known/oauth-protected-resource/mcp` -> `http://localhost:3003/.well-known/oauth-protected-resource/mcp/mcp` in MCP Inspector.

rinormaloku replied:
fixed this

rinormaloku committed:
adds support for resource server metadata endpoint and other MCP server responsibilities

rinormaloku committed:
Update dependencies and enhance resource metadata handling

- Updated `heck` dependency to version 0.5.0 in `Cargo.lock` and `Cargo.toml`.
- Refactored `ResourceMetadata` to use a flexible structure with additional properties.
- Simplified JSON construction for resource metadata in the MCP server.
- Removed deprecated fields from the schema and updated documentation accordingly.

rinormaloku committed:
Refactor Makefile targets for validation dependencies

- Renamed `run-validate-deps` to `run-validation-deps` for consistency.
- Updated corresponding commands in the `examples/mcp-authentication/README.md` to reflect the new target names.

howardjohn commented on crates/agentgateway/src/mcp/sse.rs:
```diff
@@ -233,23 +249,58 @@ pub struct McpTarget {
+		::http::Response::builder()
+			.status(StatusCode::OK)
+			.header("content-type", "application/json")
+			.header("access-control-allow-origin", "*")
+			.header("access-control-allow-methods", "GET, OPTIONS")
+			.header("access-control-allow-headers", "content-type")
```
This is a little suspicious. Shouldn't the CORS filter handle this? I believe these apply even with the MCP path

howardjohn commented on crates/agentgateway/src/mcp/sse.rs:
```diff
@@ -233,23 +249,58 @@ pub struct McpTarget {
+			.header("access-control-allow-methods", "GET, OPTIONS")
+			.header("access-control-allow-headers", "content-type")
+			.body(axum::body::Body::from(Bytes::from(
+				serde_json::to_string(&json_body).unwrap_or_default(),
```
nit: we should probably make to_string() failing return a INTERNAL_SERVER_ERROR vs the .body() call which cannot actually fail (only fails from invalid headers, and the headers are static)

howardjohn commented on crates/agentgateway/src/mcp/sse.rs:
```diff
@@ -318,15 +424,28 @@ impl App {
+		let mut upstream = client.simple_call(ureq).await?;
+
+		// Add CORS headers to the response
+		let headers = upstream.headers_mut();
```
same comment on CORS

howardjohn commented on crates/agentgateway/src/types/agent.rs:
```diff
@@ -1067,30 +1070,80 @@ pub struct A2aPolicy {}
+		// of fields we compute.
+		for (key, value) in &self.extra {
+			let snake = key.to_snake_case();
+			if snake == "resource" || snake == "authorization_servers" {
```
we add 2 other ones too. But it doesn't matter since we override them?
